openwisp_controller/connection/base/models.py

```diff
@@ -110,27 +110,35 @@ def __str__(self):
 
     def save(self, *args, **kwargs):
         super().save(*args, **kwargs)
-        self.auto_add_to_devices()
+        if self.auto_add:
+            transaction.on_commit(
+                lambda: auto_add_credentials_to_devices.delay(
+                    credential_id=self.pk, organization_id=self.organization_id
+                )
+            )
 
-    def auto_add_to_devices(self):
+    @classmethod
+    def auto_add_to_devices(cls, credential_id, organization_id):
         """
         When ``auto_add`` is ``True``, adds the credentials
         to each relevant ``Device`` and ``DeviceConnection`` objects
         """
-        if not self.auto_add:
-            return
-        device_model = load_model('config', 'Device')
-        devices = device_model.objects.exclude(config=None)
-        org = self.organization
-        if org:
-            devices = devices.filter(organization=org)
+        DeviceConnection = load_model('connection', 'DeviceConnection')
+        Device = load_model('config', 'Device')
+
+        devices = Device.objects.exclude(config=None)
+        if organization_id:
+            devices = devices.filter(organization_id=organization_id)
         # exclude devices which have been already added
-        devices = devices.exclude(deviceconnection__credentials=self)
+        devices = devices.exclude(deviceconnection__credentials_id=credential_id)
+        device_connections = []
         for device in devices:
-            DeviceConnection = load_model('connection', 'DeviceConnection')
-            conn = DeviceConnection(device=device, credentials=self, enabled=True)
+            conn = DeviceConnection(
+                device=device, credentials_id=credential_id, enabled=True
+            )
             conn.full_clean()
-            conn.save()
+            device_connections.append(conn)
+        DeviceConnection.objects.bulk_create(device_connections)
```

Comment:
what if `device_connections` is empty? Is a query generated in this case?
If it is, we should avoid it.